File: app/src/main/java/com/nearsoft/nearbooks/view/activities/BaseActivity.java
package com.nearsoft.nearbooks.view.activities;

import android.app.SearchManager;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;

import com.nearsoft.nearbooks.NearbooksApplication;
import com.nearsoft.nearbooks.R;
import com.nearsoft.nearbooks.di.components.BaseActivityComponent;
import com.nearsoft.nearbooks.di.components.DaggerBaseActivityComponent;
import com.nearsoft.nearbooks.di.components.NearbooksApplicationComponent;
import com.nearsoft.nearbooks.di.modules.BaseActivityModule;
import com.nearsoft.nearbooks.models.sqlite.User;
import com.nearsoft.nearbooks.sync.SyncChangeHandler;

import javax.inject.Inject;

import dagger.Lazy;
import rx.Subscription;
import rx.subscriptions.CompositeSubscription;

/**
 * Base activity.
 * Created by epool on 11/17/15.
 */
public abstract class BaseActivity extends AppCompatActivity {

    private final CompositeSubscription mCompositeSubscription = new CompositeSubscription();
    @Inject
    protected SyncChangeHandler mSyncChangeHandler;
    @Inject
    protected Lazy<User> mLazyUser;
    private ViewDataBinding mBinding;
    private BaseActivityComponent mBaseActivityComponent;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        initDependencies();
        super.onCreate(savedInstanceState);

        int layoutResourceId = getLayoutResourceId();

        mBinding = DataBindingUtil.setContentView(this, layoutResourceId);

        Toolbar toolbar = (Toolbar) findViewById(getToolbarId());
        if (toolbar != null) {
            setSupportActionBar(toolbar);
        }

        handleIntent(getIntent());
    }

    private void initDependencies() {
        mBaseActivityComponent = DaggerBaseActivityComponent
                .builder()
                .nearbooksApplicationComponent(getNearbooksApplicationComponent())
                .baseActivityModule(new BaseActivityModule(this))
                .build();
        injectComponent(mBaseActivityComponent);
    }

    @Override
    protected void onResume() {
        super.onResume();

        mSyncChangeHandler.onResume();
    }

    @Override
    protected void onPause() {
        mSyncChangeHandler.onPause();

        super.onPause();
    }

    @Override
    protected void onDestroy() {
        unSubscribeFromActivity();

        super.onDestroy();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        setIntent(intent);
        handleIntent(intent);
    }

    private void handleIntent(Intent intent) {
        if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
            String query = intent.getStringExtra(SearchManager.QUERY);
            if (this instanceof OnSearchListener) {
                OnSearchListener onSearchListener = (OnSearchListener) this;
                onSearchListener.onSearchRequest(query);
            }
            for (Fragment fragment : getSupportFragmentManager().getFragments()) {
                if (fragment instanceof OnSearchListener) {
                    OnSearchListener onSearchListener = (OnSearchListener) fragment;
                    onSearchListener.onSearchRequest(query);
                }
            }
        }
    }

    protected abstract int getLayoutResourceId();

    protected int getToolbarId() {
        return R.id.toolbar;
    }

    protected <T extends ViewDataBinding> T getBinding(Class<T> clazz) {
        return clazz.cast(mBinding);
    }

    protected void injectComponent(BaseActivityComponent baseActivityComponent) {
        baseActivityComponent.inject(this);
    }

    protected void subscribeToActivity(Subscription subscription) {
        mCompositeSubscription.add(subscription);
    }

    protected void unSubscribeFromActivity() {
        mCompositeSubscription.clear();
    }

    public BaseActivityComponent getBaseActivityComponent() {
        return mBaseActivityComponent;
    }

    public SyncChangeHandler getSyncChangeHandler() {
        return mSyncChangeHandler;
    }

    protected NearbooksApplicationComponent getNearbooksApplicationComponent() {
        return NearbooksApplication.getNearbooksApplicationComponent();
    }

    protected BaseActivityModule getBaseActivityModule() {
        return new BaseActivityModule(this);
    }

    public interface OnSearchListener {

        void onSearchRequest(String query);

    }

}

File: app/src/main/java/com/nearsoft/nearbooks/view/activities/BaseActivity.kt
package com.nearsoft.nearbooks.view.activities

import android.app.SearchManager
import android.content.Intent
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import com.nearsoft.nearbooks.NearbooksApplication
import com.nearsoft.nearbooks.R
import com.nearsoft.nearbooks.di.components.BaseActivityComponent
import com.nearsoft.nearbooks.di.components.DaggerBaseActivityComponent
import com.nearsoft.nearbooks.di.modules.BaseActivityModule
import com.nearsoft.nearbooks.models.view.User
import com.nearsoft.nearbooks.sync.SyncChangeHandler
import dagger.Lazy
import rx.Subscription
import rx.subscriptions.CompositeSubscription
import javax.inject.Inject

/**
 * Base activity.
 * Created by epool on 11/17/15.
 */
abstract class BaseActivity : AppCompatActivity() {

    private val mCompositeSubscription = CompositeSubscription()
    @Inject
    lateinit var syncChangeHandler: SyncChangeHandler
    @Inject
    lateinit protected var mLazyUser: Lazy<User>
    private lateinit var mBinding: ViewDataBinding
    val baseActivityComponent: BaseActivityComponent by lazy {
        DaggerBaseActivityComponent
                .builder()
                .nearbooksApplicationComponent(NearbooksApplication.applicationComponent)
                .baseActivityModule(BaseActivityModule(this))
                .build()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        initDependencies()
        super.onCreate(savedInstanceState)

        val layoutResourceId = getLayoutResourceId()

        mBinding = DataBindingUtil.setContentView<ViewDataBinding>(this, layoutResourceId)

        val toolbar = findViewById(getToolbarId()) as Toolbar?
        setSupportActionBar(toolbar)

        handleIntent(intent)
    }

    private fun initDependencies() {
        injectComponent(baseActivityComponent)
    }

    override fun onResume() {
        super.onResume()

        syncChangeHandler.onResume()
    }

    override fun onPause() {
        syncChangeHandler.onPause()

        super.onPause()
    }

    override fun onDestroy() {
        unSubscribeFromActivity()

        super.onDestroy()
    }

    override fun onNewIntent(intent: Intent) {
        setIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent) {
        if (Intent.ACTION_SEARCH == intent.action) {
            val query = intent.getStringExtra(SearchManager.QUERY)
            if (this is OnSearchListener) {
                this.onSearchRequest(query)
            }
            for (fragment in supportFragmentManager.fragments) {
                if (fragment is OnSearchListener) {
                    fragment.onSearchRequest(query)
                }
            }
        }
    }

    protected abstract fun getLayoutResourceId(): Int

    protected fun getToolbarId(): Int = R.id.toolbar

    protected fun <T : ViewDataBinding> getBinding(clazz: Class<T>): T {
        return clazz.cast(mBinding)
    }

    protected open fun injectComponent(baseActivityComponent: BaseActivityComponent) {
        baseActivityComponent.inject(this)
    }

    override fun setSupportActionBar(toolbar: Toolbar?) {
        if (toolbar != null) {
            super.setSupportActionBar(toolbar)
        }
    }

    protected fun subscribeToActivity(subscription: Subscription) {
        mCompositeSubscription.add(subscription)
    }

    protected fun unSubscribeFromActivity() {
        mCompositeSubscription.clear()
    }

    protected val baseActivityModule: BaseActivityModule
        get() = BaseActivityModule(this)

    interface OnSearchListener {

        fun onSearchRequest(query: String)

    }

}
